platform/linux-generic/test/validation/api/ml/ml_linux.c

```diff
@@ -0,0 +1,1173 @@
+	CU_ASSERT_FATAL(ret == 0);
+
+	/* When odp_ml_model_unload_start() succeeded, continue to check completion
+	 * status */
+	for (int i = 0; i < TIMEOUT; i++) {
+		ret = odp_ml_model_unload_status(global.ml_model, 0, &result);
+		if (ret)
+			break;
+
+		/* ret = 0 meaning run has not finished, continue to check status */
+		odp_time_wait_ns(wait_ns);
+	}
+
+	CU_ASSERT_FATAL(ret > 0);
+	CU_ASSERT_EQUAL(result.error_code, 0);
+	CU_ASSERT_PTR_EQUAL(result.user_ptr, user_ptr);
+
+	/* odp_ml_model_unload does not modify data in user_ptr */
+	if (result.user_ptr)
+		CU_ASSERT_EQUAL(*(int *)result.user_ptr, dummy);
+}
+
+static int
+get_result_from_ml_compl_event(odp_ml_load_result_t *load_result, odp_ml_run_result_t *run_result)
+{
+	int ret;
+	odp_event_t ev;
+	odp_ml_compl_t compl;
+	odp_event_type_t ev_type;
+	odp_queue_t from_queue = ODP_QUEUE_INVALID;
+	uint64_t sched_wait = odp_schedule_wait_time(global.wait_ns);
+
+	/* Run event scheduler to find the ml completion event */
+	for (int i = 0; i < TIMEOUT; i++) {
+		ev = odp_schedule(&from_queue, sched_wait);
+		if (ev != ODP_EVENT_INVALID)
+			break;
+	}
+
+	CU_ASSERT_NOT_EQUAL(ev, ODP_EVENT_INVALID);
+	if (ev == ODP_EVENT_INVALID)
+		return -1;
+
+	ev_type = odp_event_type(ev);
+	CU_ASSERT_EQUAL(from_queue, global.queue);
+	CU_ASSERT_EQUAL(ev_type, ODP_EVENT_ML_COMPL);
+	if (from_queue != global.queue || ev_type != ODP_EVENT_ML_COMPL) {
+		odp_event_free(ev);
+		return -1;
+	}
+
+	compl = odp_ml_compl_from_event(ev);
+	CU_ASSERT_NOT_EQUAL(compl, ODP_ML_COMPL_INVALID);
+
+	if (load_result) {
+		CU_ASSERT(odp_ml_compl_load_result(compl, NULL) == 0);
+		ret = odp_ml_compl_load_result(compl, load_result);
+	} else {
+		CU_ASSERT(odp_ml_compl_run_result(compl, NULL) == 0);
+		ret = odp_ml_compl_run_result(compl, run_result);
+	}
+
+	CU_ASSERT_EQUAL(ret, 0);
+	odp_ml_compl_free(compl);
+
+	return ret;
+}
+
+/* Test asynchronous model loading in ODP_ML_COMPL_MODE_EVENT mode */
+static void test_ml_model_load_async_event(void)
+{
+	int ret;
+	odp_ml_compl_t compl;
+	odp_ml_load_result_t result;
+	odp_ml_compl_param_t compl_param;
+	int dummy = 6;
+	void *user_ptr = &dummy;
+
+	compl = odp_ml_compl_alloc(global.compl_pool);
+	CU_ASSERT_PTR_NOT_EQUAL_FATAL(compl, ODP_ML_COMPL_INVALID);
+
+	odp_ml_compl_param_init(&compl_param);
+	compl_param.mode = ODP_ML_COMPL_MODE_EVENT;
+	compl_param.event = odp_ml_compl_to_event(compl);
+	compl_param.queue = global.queue;
+	compl_param.user_ptr = user_ptr;
+
+	ret = odp_ml_model_load_start(global.ml_model, &compl_param);
+	CU_ASSERT_EQUAL(ret, 0);
+
+	/* Return when odp_ml_model_load_start() failed */
+	if (ret) {
+		odp_ml_compl_free(compl);
+		ODPH_ERR("ML model odp_ml_model_load_start() failed\n");
+		return;
+	}
+
+	/* Run event scheduler to find the ml completion event and verify it */
+	if (get_result_from_ml_compl_event(&result, NULL)) {
+		ret = odp_ml_model_unload(global.ml_model, NULL);
+		CU_ASSERT_EQUAL(ret, 0);
```

Comment:
Sync unload call needs to be removed from async test.

get_result_from_ml_compl_event() could use ODPH_ERR() to print error reason for easier debug when this assert happens.